doc/source/user/basics.interoperability.rst

```diff
@@ -0,0 +1,352 @@
+re-use existing code with minimal (ideally zero) porting overhead. With this
+goal in mind, various protocols are defined for implementations of
+multi-dimensional arrays with high-level APIs matching NumPy.
+
+Using arbitrary objects in NumPy
+--------------------------------
+
+When NumPy functions encounter a foreign object, they will try (in order):
+
+1. The buffer protocol, described `in the Python C-API documentation
+   <https://docs.python.org/3/c-api/buffer.html>`__.
+2. The ``__array_interface__`` protocol, described
+   :ref:`in this page <arrays.interface>`. A precursor to Python’s buffer
+   protocol, it defines a way to access the contents of a NumPy array from other
+   C extensions.
+3. The ``__array__`` protocol, which asks an arbitrary object to convert itself
+   into an array.
+
+For both the buffer and the ``__array_interface__`` protocols, the object
+describes its memory layout and NumPy does everything else (zero-copy if
+possible). If that’s not possible, the object itself is responsible for
+returning a ``ndarray`` from ``__array__()``.
+
+The array interface
+~~~~~~~~~~~~~~~~~~~
+
+The :ref:`array interface <arrays.interface>` defines a protocol for array-like
+objects to re-use each other’s data buffers. Its implementation relies on the
+existence of the following attributes or methods:
+
+-  ``__array_interface__``: a Python dictionary containing the shape, the
+   element type, and optionally, the data buffer address and the strides of an
+   array-like object;
+-  ``__array__()``: a method returning the NumPy ndarray view of an array-like
+   object;
+-  ``__array_struct__``: a ``PyCapsule`` containing a pointer to a
+   ``PyArrayInterface`` C-structure.
+
+The ``__array_interface__`` and ``__array_struct__`` attributes can be inspected
+directly:
+
+ >>> import numpy as np
+ >>> x = np.array([1, 2, 5.0, 8])
+ >>> x.__array_interface__
+ {'data': (94708397920832, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (4,), 'version': 3}
+ >>> x.__array_struct__
+ <capsule object NULL at 0x7f798800be40>
+
+The ``__array_interface__`` attribute can also be used to manipulate the object
+data in place:
+
+ >>> class wrapper():
+ ...     pass
+ ... 
+ >>> arr = np.array([1, 2, 3, 4])
+ >>> buf = arr.__array_interface__
+ >>> buf
+ {'data': (140497590272032, False), 'strides': None, 'descr': [('', '<i8')], 'typestr': '<i8', 'shape': (4,), 'version': 3}
+ >>> buf['shape'] = (2, 2)
+ >>> w = wrapper()
+ >>> w.__array_interface__ = buf
+ >>> new_arr = np.array(w, copy=False)
+ >>> new_arr
+ array([[1, 2],
+        [3, 4]])
+
+We can check that ``arr`` and ``new_arr`` share the same data buffer:
+
+ >>> new_arr[0, 0] = 1000
+ >>> new_arr
+ array([[1000,    2],
+        [   3,    4]])
+ >>> arr
+ array([1000, 2, 3, 4])
+
+
+The ``__array__`` protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``__array__`` protocol acts as a dispatch mechanism and ensures that any
```

Comment:
It's not a dispatch mechanism I'd say. Dispatching means something like "deferring execution to an implementation elsewhere". That is not the case here. Instead let's just say it's a way to request the object to turn itself into a `np.ndarray` instance. The original object, and any attributes/behavior it may have had, is lost.

Reply:
OK - I rephrased it, let me know if it reads better.